src/Resources/skeleton/forgottenPassword/ForgottenPasswordController.tpl.php

```diff
@@ -0,0 +1,112 @@
+<?= "<?php\n" ?>
+
+namespace <?= $namespace; ?>;
+
+use <?= $user_full_class_name ?>;
+use <?= $token_full_class_name ?>;
+use <?= $request_form_full_class_name ?>;
+use <?= $resetting_form_full_class_name ?>;
+use Symfony\Bundle\FrameworkBundle\Controller\<?= $parent_class_name; ?>;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
+
+class <?= $class_name; ?> extends <?= $parent_class_name; ?><?= "\n" ?>
+{
+    /**
+     * @Route("/forgotten-password", name="app_forgotten_password")
+     */
+    public function request(Request $request, \Swift_Mailer $mailer): Response
+    {
+        $form = $this->createForm(<?= $request_form_class_name ?>::class);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            $user = $this->getDoctrine()->getRepository(<?= $user_class_name ?>::class)->findOneBy([
+                '<?= $email_field ?>' => $form->get('<?= $email_field ?>')->getData(),
+            ]);
+
+            // do not reveal whether an user account was found or not
+            if (!$user) {
+                return $this->redirectToRoute('app_check_email');
+            }
+
+            // generate a reset password token, that the user could use to change his password
+            $resetPasswordToken = new <?= $token_class_name?>($user);
+            $this->getDoctrine()->getManager()->persist($resetPasswordToken);
+
+            $message = (new \Swift_Message('Your password reset request'))
+                ->setFrom(['noreply@mydomain.com' => 'Noreply'])
+                ->setTo($user-><?= $email_getter ?>())
+                ->setBody($this->renderView('forgottenPassword/email.txt.twig', [
+                    'token' => $resetPasswordToken,
+                ]))
+            ;
+            $mailer->send($message);
+
+            $this->getDoctrine()->getManager()->flush();
+
+            return $this->redirectToRoute('app_check_email');
+        }
+
+        return $this->render('forgottenPassword/request.html.twig', [
+            'requestForm' => $form->createView(),
+        ]);
+    }
+
+    /**
+     * @Route("/check-email", name="app_check_email")
+     */
+    public function checkEmail()
+    {
+        return $this->render('forgottenPassword/check_email.html.twig', [
```

Comment:
This route is always available, so you open it even if you did not request password resetting. What about to check referer for it? Something like:
```php
        // Redirect if user does not come from request reset password page
        $refererUrl = $request->headers->get('referer');
        $requestUrl = $this->generateUrl('app_forgotten_password');
        if (!$refererUrl) {
            return $this->redirect($requestUrl);
        }
        // It's assumed we didn't have any query params on the Request page,
        // otherwise we need a more clever solution here like using parse_url()
        if ($requestUrl !== substr($refererUrl, -strlen($requestUrl))) {
            return $this->redirect($requestUrl);
        }

       return $this->render(...)
```

So, only users who were redirected from `app_forgotten_password` would be able to see this page

Reply:
A referer check is pretty weak. What about adding some extra (random) token both in User session & as an attribute to `checkEmail` route?

Reply:
Sounds cool, I like it 👍 So, when "check email" page is rendered - we need to invalidate that token (remove it), right? And if the token is invalid - redirect to `app_forgotten_password`

Reply:
Exactly!